### ISSUE TITLE
Editor Loses Focus on Tab – Must Click to Type

### ISSUE THREAD

**Issue by babalugats76** [ianstormtaylor/slate#3696]
#### Do you want to request a _feature_ or report a _bug_?

Bug?

#### What's the current behavior?

When tabbing into Slate's editor, the **focus appears and then disappears**.

![tab-focus-issue](https://user-images.githubusercontent.com/6919889/82607577-4851fb80-9b76-11ea-8a62-43a4c70c9d68.gif)

Recreate using this minimally viable example: [CodeSandbox - Slate 0.58.1 - Tab Focus Issue](https://codesandbox.io/s/slate-tab-between-editor-issue-86zkl?file=/src/RichText.js)

1. Set focus to first editor
2. Type _"Mama said knock you out"_
3. Tab (to second editor)
4. Try #2

You should **not be able to immediately type into the second editor**; the focus will appear to be "robbed" from it and you will need to click inside in order to enter content.

Slate: 0.58.1
Browser: Chrome
OS: Windows

#### What's the expected behavior?

Should work much like its plain ole HTML cousin: [Tab Focus Issue - Pure HTML](https://codepen.io/babalugats76/pen/Vwvbwym)

Upon tabbing in, focus should be preserved, **allowing the end user to type immediately into the editor**, i.e., **without clicking** into it first!

**Comment by babalugats76** [comment]
Seems suspiciously similar to [#3634](https://github.com/ianstormtaylor/slate/issues/3634) (if not a duplicate).

In both cases:

- Multiple editors are involved
- The first editor works as expected
- If the first editor has previously received focus, subsequent editors require explicit clicking or multiple `focus()` calls to allow typing

**Comment by sebastianseilund** [comment]
@babalugats76 I figured something out!

**TL;DR; Workaround: Make sure that your `onChange` handler does not cause a re-render if the Slate value has not changed.**

I've been struggling with a similar issue. Switching from one editor to another, results in the second editor to be focused but with no selection/cursor (user has to click it again to actually be able to type). It doesn't happen between all editors though, which was puzzling.

I found a way to make your example work. [See this CodeSandbox](https://codesandbox.io/s/slate-tab-between-editor-issue-forked-etn6s?file=/src/TabIssue.js)

All I changed was making your `handleEditorChange` function return `prevState` if the value didn't change:

```diff,js
  const handleEditorChange = (value, field) => {
    setState((prevState) => {
+      if (prevState[field].value === value) {
+        return prevState;
+      }
      return {
        ...prevState,
        [field]: {
          ...prevState[field],
          value: value
        }
      };
    });
  };
```

Your example's `handleEditorChange` would always return a new state object, which makes the `<TabIssue>` component re-render even if the value didn't change. With my patch above, it will not re-render since the same state object is returned (React knows it doesn't need to re-render).

I made 2 even simpler examples to demonstrate the issue:

In [Example 1](https://codesandbox.io/s/slate-tab-between-editor-issue-forked-z23tr?file=/src/App.js) tabbing/clicking between the 2 editors works fine. It uses a state field for the root children array:

```js
const [value1, setValue1] = useState([
  { type: "paragraph", children: [{ text: "" }] }
]);
// ...
<Slate editor={editor1} onChange={setValue1} value={value1}>
// ...
```

Notice that when `onChange`/`setValue1` is called, React will only re-render the `<App>` component if the children array actually changed.

[Example 2](https://codesandbox.io/s/slate-tab-between-editor-issue-forked-2z70i) has the same issue as your example. It wraps the root children array in an object:

```js
const [value1, setValue1] = useState({
  wrapped: [{ type: "paragraph", children: [{ text: "" }] }]
});
// ...
<Slate
  editor={editor1}
  onChange={(wrapped) => setValue1({ wrapped })}
  value={value1.wrapped}
>
// ...
```

Notice that when `onChange` is called, React will _always_ re-render the `<App>` component even if nothing changed, since it sets a new `{wrapped}` object.

I believe the issue occurs because blurring an editor also triggers it's `onChange` before the `onChange` is also triggered for the next editor when it gets focused. This means when editor 1 is focused and you tab (or click) to editor 2, then editor 1's state is first changed, which causes a re-render, and that somehow trips up Slate's ability to select anything in the second editor.

Not sure why... I'm looking through the Slate codebase to see if there's a possible fix so we don't need this workaround.

**Comment by sebastianseilund** [comment]
The issue happens even with just 1 editor: If the `<Slate>` component's parent component is re-rendered in the same "tick"(?) as it gets focused. [CodeSandbox example with just 1 editor](https://codesandbox.io/s/slate-focus-re-render-issue-l48iz?file=/src/App.js)

**Comment by iSplasher** [comment]
@sebastianseilund 
Not sure about chromium but that fix doesn't work in firefox

**Comment by WenWenBo** [comment]
It is the default behavior of Tab, you can prevent in `onKyedown `handler

```
const onKeyDown = (e) => {
    if (e.key === "Tab") {
      e.preventDefault();
    }
  };
<Slate
     ......
      >
        <Editable
          onKeyDown={onKeyDown}
          ......
        />
      </Slate>
```
